crates/ruff_cli/src/lib.rs

```diff
@@ -52,6 +53,38 @@ enum ChangeKind {
     SourceFile,
 }
 
+enum OutputWriter {
+    Stdout(BufWriter<io::Stdout>),
+    File(BufWriter<File>),
+}
+
+impl TryFrom<&Arguments> for OutputWriter {
+    type Error = anyhow::Error;
+
+    fn try_from(args: &Arguments) -> Result<Self> {
+        Ok(match args.output_file.as_ref() {
+            Some(path) if !args.watch => OutputWriter::File(BufWriter::new(File::create(path)?)),
+            _ => OutputWriter::Stdout(BufWriter::new(io::stdout())),
+        })
+    }
+}
+
+impl Write for OutputWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self {
+            OutputWriter::Stdout(stdout) => stdout.write(buf),
+            OutputWriter::File(file) => file.write(buf),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        match self {
+            OutputWriter::Stdout(stdout) => stdout.flush(),
+            OutputWriter::File(file) => file.flush(),
+        }
+    }
+}
```

Comment:
> Do I understand the change correctly that you decided to remove the lock (I'm fine with that)?

Yes.

> I think this solution is overkill without the lock. Instead, we can pass a `&dyn Writer` to `Printer` because `Writer` already exposes all methods we need. Changing Printer to accept a `&dyn Writer` avoids Rust generating two implementations of Printer, one for a file writer and one for the stdout writer (that would result in a lot of duplicated code). I doubt that this creates a significant performance hit.

Yes, that makes sense. I'll update it.